### ISSUE TITLE
Proto Extension does not use tab_size for formatting Protobuf files

### ISSUE THREAD

**Issue by cemoktra** [zed-industries/zed#20009]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

The following configuration is ignored in latest version when language server is not running:
```json
  "languages": {
    "Proto": {
      "tab_size": 2
    }
  }
```

### Environment

Zed: v0.159.5 (Zed)
OS: Linux X11 ubuntu 24.04
Memory: 31.1 GiB
Architecture: x86_64
GPU: Intel(R) UHD Graphics (TGL GT1) || Intel open-source Mesa driver || Mesa 24.0.9-0ubuntu0.2

### If applicable, add mockups / screenshots to help explain present your vision of the feature

Pressing tab adds 4 spaces despite 2 being configured

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary><pre>

</pre></details>

**Comment by SomeoneToIgnore** [comment]
Do you have an `.editorconfig` in that repo?

**Comment by cemoktra** [comment]
no

**Comment by notpeter** [comment]
1. Do you have the "Proto" extension installed? 

2. Is the file being identified as a `Proto` language file and setting the Language mode appropriately?
<img width="267" alt="Screenshot 2024-10-31 at 13 59 33" src="https://github.com/user-attachments/assets/3eaa26a9-2a26-4cd2-878a-98a1a86aaa48">

3. What do you mean when you say "language server is not running" (e.g. is the extension uninstalled? Or...). Is there a setting you can change which causes indent to work properly (e.g. running the language server)?

**Comment by cemoktra** [comment]
Yes i have the proto extension installed. It used to work without installing/running the proto language server manually. But not i saw that in the logs zed was complaining about not finding language server

**Comment by cemoktra** [comment]
I updated to newest version today. The problem still exists. Additionally when i misformat the file and select formatting, it won't fix it. It does not produce any trace/log messages.

I'm using language server [protols](https://github.com/coder3101/protols)

**Comment by notpeter** [comment]
Protols uses clang-format under the hood. Can you confirm you have `clang-format` installed?

I couldn't actually get protols working at all, so as a workaround I am invoking `clang-format` directly and it works for me:
```
    "Proto": {
      "tab_size": 4,
      "formatter": {
        "external": {
          "command": "clang-format",
          "arguments": ["-style={IndentWidth: 4, ColumnLimit: 0}"]
        }
      }
    },
```
Alternatively, you can create a `.clang-format` with your [clang-format style options](https://clang.llvm.org/docs/ClangFormatStyleOptions.html) instead of providing them as CLI arguments.

**Comment by cemoktra** [comment]
Ok, well i think i found a ocompatible clang-format config. Anyway it used to work and suddenly stopped working without messing with the config

**Comment by notpeter** [comment]
> Our CI uses https://buf.build/ and i'm not able to configure it as external formatter, also it is very slow and times out.

`buf format` does not support formatting handling single files via stdin yet.
There is an upstream issue here:

- https://github.com/bufbuild/buf/issues/1035

**Comment by cemoktra** [comment]
`"-style={IndentWidth: 2, ColumnLimit: 0, Language: Proto, BasedOnStyle: Google}"`

**Comment by beniaminzagan** [comment]
@cemoktra is this something that still an issue at this date @cemoktra or did you manage to get it running? @notpeter did you want to add this to the doc's - maybe we should wait and see if its still an issue if so we can add it? As its from 4 months ago.

**Comment by cemoktra** [comment]
No currently it's working

**Comment by beniaminzagan** [comment]
@cemoktra glad to hear, i will be closing this issue if you have any others questions / issues feel free to open a new one! Happy coding!

**Comment by notpeter** [comment]
Docs were never updated because `buildbuf/buf` did not support being invoked a pre-save formatter where the file content is passed via stdin / filename is passed as argument as is needed for custom formatters in Zed.

As of an hour ago (2026-09-15) this has been fixed upstream in main, but it's not yet available in a tagged release.  Buf tends to do a minor release once a month and patch releases only for regressions fixes, so likely it'll be sometime in October.

Once that's available in a stable `buf` release something like this should be possible (untested): 

```jsonc
  "languages": {
    "Proto":
      "formatter": {
        "external": {
          "command": "buf",
          "arguments": ["format", "--stdin-filepath", "{buffer_path}"]
        }
      }
    }
  }
```

See: 
- https://github.com/bufbuild/buf/issues/1035
- https://github.com/bufbuild/buf/pull/4681

P.S. I no longer work at Zed or use Protobuf in $dayjob so I won't be validating this or submitting a docs PR, but figured other folks would be happy to see the situation improve nearly 22months after initial filing.
